extensions/amp-lightbox-viewer/0.1/service/lightbox-manager-impl.js

```diff
@@ -120,20 +134,45 @@ export class LightboxManager {
    * @private
    */
   processLightboxElement_(element) {
-    if (element.tagName == 'AMP-CAROUSEL') {
-      // TODO: special case carousel
+    if (element.tagName.toLowerCase() == CAROUSEL_TAG) {
+      const lightboxGroupId = element.getAttribute('lightbox') ||
+       'carousel' + (element.getAttribute('id') || this.counter_++);
+      this.getSlidesFromCarousel_(element).then(slides => {
+        slides.forEach(slide => {
+          // TODO: review naming conventions for component attributes
+          if (!slide.hasAttribute('lightbox-exclude')) {
+            slide.setAttribute('lightbox', lightboxGroupId);
+            this.processBaseLightboxElement_(slide, lightboxGroupId);
+          }
+        });
+      });
     } else {
       const lightboxGroupId = element.getAttribute('lightbox') || 'default';
-      if (!this.lightboxGroups_[lightboxGroupId]) {
-        this.lightboxGroups_[lightboxGroupId] = [];
-      }
-      this.lightboxGroups_[lightboxGroupId]
-          .push(dev().assertElement(element));
-      if (this.meetsHeuristicsForTap_(element)) {
-        const viewer = elementByTag(this.ampdoc_.getRootNode(), VIEWER_TAG).id;
-        element.setAttribute('on', 'tap:' + viewer.id + '.activate');
-      }
+      this.processBaseLightboxElement_(element, lightboxGroupId);
+    }
+  }
+
+  processBaseLightboxElement_(element, lightboxGroupId) {
+    if (!this.lightboxGroups_[lightboxGroupId]) {
+      this.lightboxGroups_[lightboxGroupId] = [];
     }
+    this.lightboxGroups_[lightboxGroupId]
+        .push(dev().assertElement(element));
+    if (this.meetsHeuristicsForTap_(element)) {
+      const viewer = elementByTag(this.ampdoc_.getRootNode(), VIEWER_TAG);
+      element.setAttribute('on', `tap:${viewer.id}.activate`);
+    }
+  }
+
+  /**
+   * @param {!Element} element
+   * @return {!Promise<!Array<!Element>>}
+   * @private
+   */
+  getSlidesFromCarousel_(element) {
+    return element.signals().whenSignal(CommonSignals.LOAD_END).then(() => {
+      return toArray(scopedQuerySelectorAll(element, SLIDE_SELECTOR));
```

Comment:
This is fine but note that `element.querySelectorAll('.someClassName)` will still only return `someClassName` descendants of `element`.

The whole thing is pretty confusing and something I learned recently. **When querySelector goes wrong** section of this article https://developers.google.com/web/updates/2013/03/What-s-the-CSS-scope-pseudo-class-for has a good explanation of what `:scope` pseudo-selector (which our `scopedQuerySelectorAll` is based on) is and what issues it fixes, it is worth a quick read.

Reply:
Yes, I think the intention is to return `someClassName` descendants of `element`. The `element` here is the carousel that I'm lightboxing. I'm looking for the slides of that carousel only, which are necessarily descendants of the lightbox. If I don't scope this selector, I might accidentally get slides of other carousels in the document. So I think this implementation is consistent with the intention? 

Maybe what's confusing here is the naming of the input parameter. It's currently called `element`, but should really be `carousel`.

Reply:
I think `querySelector` returns only one element, whereas `querySelectorAll` returns all elements that match the selector. We do want `querySelectorAll`. As to why not use `element.querySelectorAll`, I think that's precisely due to the issue in the article. I think the usage of `scopedQuerySelectorAll` here is the correct one (the other ones also don't lint =P ).

Reply:
Oh I see. There's a performance hit. Basically `element.querySelectorAll` will still work, but won't have the perf hit associated with `scopedQuerySelectorAll`. Since this use case isn't necessary for `scopedQuerySelectorAll`, it's better not to have the perf it and just to `/*OK*/` lint. Is that what you mean @aghassemi ?

Reply:
correct. The reason I brought it up wasn't necessarily for the perf gains (which are not much if any), but mostly to tell what I learned regarding scropeSelector recently which was unintuitive to me so though it is worth sharing.

Alss note that the reasons our presubmit complains about `element.querySelectorAll(VARIABLE)` is that a `variable` is passed to it so it can not make a the right judgment whether the query is one of the problematic ones or not. If a string literal was passed to it, it would not have complained.